src/reference/add-wasm-support-to-crate.md

```diff
@@ -0,0 +1,156 @@
+# How to Add WebAssembly Support to a General-Purpose Crate
+
+This section is for general-purpose crate authors who want to support
+WebAssembly.
+
+## Maybe Your Crate Already Supports WebAssembly!
+
+Review the information about [what kinds of things can make a general-purpose
+crate *not* portable for WebAssembly](./which-crates-work-with-wasm.html). If
+your crate doesn't have any of those things, it likely already supports
+WebAssembly!
+
+You can always check by running `cargo build` for the WebAssembly target:
+
+```
+cargo build --target wasm32-unknown-unknown
+```
+
+If that command fails, then your crate doesn't support WebAssembly right now. If
+it doesn't fail, then your crate *might* support WebAssembly. You can be 100%
+sure that it does (and continues to do so!) by [adding tests for wasm and
+running those tests in CI.](#maintaining-ongoing-support-for-webassembly)
+
+## Adding Support for WebAssembly
+
+### Avoid Performing I/O Directly
+
+On the Web, I/O is always asynchronous, and there isn't a file system. Factor
+I/O out of your library, let users perform the I/O and then pass the input
+slices to your library instead.
+
+For example, refactor this:
+
+```rust
+use std::fs;
+use std::path::Path;
+
+pub fn parse_thing(path: &Path) -> Result<MyThing, MyError> {
+    let contents = fs::read(path)?;
+    // ...
+}
+```
+
+Into this:
+
+```rust
+pub fn parse_thing(contents: &[u8]) -> Result<MyThing, MyError> {
+    // ...
+}
+```
+
+### Add `wasm-bindgen` as a Dependency
+
+If you need to interact with the outside world (i.e. you can't have library
+consumers drive that interaction for you) then you'll need to add `wasm-bindgen`
+(and `js-sys` and `web-sys` if you need them) as a dependency for when
+compilation is targeting WebAssembly:
+
+```toml
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+wasm-bindgen = "0.2"
+js-sys = "0.3"
+web-sys = "0.3"
+```
```

Comment:
It'd be awesome if we had an example crate to add here, something like adding wasm support to an existing crate. The biggest one that comes to mind is `rand`, although due to the optional nature of the support it's not an ideal example per se

Reply:
Yes, it would be. Happy to add any you think of in follow ups.